### PR TITLE
Catlin: Validate images used in task steps are tagged

### PR DESCRIPTION
**Pull request by SM43**
Signed-off-by: Shivam Mukhade <smukhade@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

File: catlin/pkg/validator/kind_validator.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+func NewKindValidator(cli app.CLI, r *parser.Resource) Validator {
+
+	switch r.Kind {
+	case "Task":
+		return &TaskValidator{res: r, log: cli.Logger().Sugar()}
+	}
+
+	return nil
```
Comment (sthaha) [inline catlin/pkg/validator/kind_validator.go:29]: this should return an error `UnknownKindError` instead of nil which can be checked by the caller before adding to the list of `validators`

File: catlin/pkg/validator/kind_validator.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// limitations under the License.
+
+package validator
+
+import (
+	"github.com/tektoncd/plumbing/catlin/pkg/app"
+	"github.com/tektoncd/plumbing/catlin/pkg/parser"
+)
+
+func NewKindValidator(cli app.CLI, r *parser.Resource) Validator {
```
Comment (sthaha) [inline catlin/pkg/validator/kind_validator.go:22]: How about `ValidatorForKind` that returns `Validator, error` ?

File: catlin/pkg/validator/task_validator.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+var _ Validator = (*TaskValidator)(nil)
+
+func (t *TaskValidator) Validate() Result {
+
+	result := Result{}
+
+	var task *v1beta1.Task
+	str, _ := json.Marshal(t.res.Unstructured.Object)
```
Comment (sthaha) [inline catlin/pkg/validator/task_validator.go:39]: I think it may be better to add a function to the `ParsedResource` to convert to the right type. 
```go
var task v1beta.Task
err := r.ToType(&task)
```
and then 

```
for _, step := ... 
```

This makes it quite easy to add  support for other types as well. Please refer to the code in parse to properly unmarshall to a typed object.

Comment (SM43) [reply catlin/pkg/validator/task_validator.go]: Thank you :)

File: catlin/pkg/validator/task_validator.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		result.Error("failed to decode task - %s", err)
+		return result
+	}
+
+	for _, step := range res.(*v1beta1.Task).Spec.Steps {
+		img := strings.FieldsFunc(step.Image, func(c rune) bool { return c == ':' })
+
+		if len(img) == 1 || img[1] == "latest" {
+			result.Error("Task image (%s) must be tagged with a specific version", step.Image)
+		}
```
Comment (vdemeester) [inline catlin/pkg/validator/task_validator.go:48]: We should use a package that parse image reference, like `go-containerregistry` `pkg/name` package or docker's packages. This doesn't take digests into account for example (although this code would work)

File: catlin/pkg/cmd/validate/validate.go [hunk 5]
```diff
@@ -76,6 +76,11 @@ func validate(cli app.CLI, path string) error {
 		validator.NewContentValidator(cli, res),
 	}
 
+	kindValidator, err := validator.ValidatorForKind(cli, res)
```
Comment (sthaha) [inline catlin/pkg/cmd/validate/validate.go:79]: nit: 
```go
if v, err := validator.ForKind(cli, res.Kind); err == nil {
  validators = append(validators, v)
}
```

On a second thought, it seems like an anti-pattern to ignore error. How about a `NoopValidator` if kind has no specific validator and `err` will a `not nil` if there was a failure to create validator at which point the CLI should exit with that error? 
So .. the code would look like 

```go 
v, err := validator.ForKind(res.Kind)
if err != nil {
  return err
}
validators = append(validators, v)

```

File: catlin/pkg/validator/task_validator.go (hunk 4 above)
Comment (sthaha) [reply catlin/pkg/validator/task_validator.go:48]: Good point! I didn't think of it at all! 🤗

Comment (SM43) [reply catlin/pkg/validator/task_validator.go:48]: `digest` would not be mandatory right? the image should be tagged with a version and not `latest`. Is that correct?

Comment (vdemeester) [reply catlin/pkg/validator/task_validator.go:48]: @SM43 that's correct. Using a digest, `foo@sha256:abc…` would work too. I think, the current code "works" with digests, but *by change*. Although it's not gonna change in the long-term future, using already existing like to parse/extra part of the image reference is a long-term viable thing to do (allow to validate more accurately the image reference, etc.)

File: catlin/pkg/cmd/validate/validate.go (hunk 5 above)
Comment (SM43) [reply catlin/pkg/cmd/validate/validate.go]: When do we want `ForKind` to return an error? we are just checking the kind and returning a validator if exists otherwise we return `NoopValidator`. we aren't doing anything on the resource inside as it is already parsed. so, not sure when we should return an error :thinking:
If res is of  `UnknownKind` the parser would fail and return from there.

File: catlin/pkg/validator/testdata.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    tekton.dev/displayName: My Example Task
+spec:
+  description: |-
+    A summary of the resource
+
+    A para about this valid task
+
+  steps:
+  - name: hello
+    image: ubuntu
```
Comment (sthaha) [inline catlin/pkg/validator/testdata.go:50]: @vdemeester @chmouel is this valid? shouldn't we insist on a tag ? 
cc: @SM43

File: catlin/pkg/validator/testdata.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  - name: foo-bar
+    image: abc.io/fedora:1.0@sha256:deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f
+`
+
+	taskWithInvalidImageRef = `
+---
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: valid
```
Comment (sthaha) [inline catlin/pkg/validator/testdata.go:35]: this need to be called `invalid` 😂

File: catlin/pkg/validator/testdata.go (hunk 6 above)
Comment (sthaha) [reply catlin/pkg/validator/testdata.go:50]: this will fail validation :D 
I went by the name of the task - `valid`

Comment (vdemeester) [reply catlin/pkg/validator/testdata.go:50]: This would be invalid indeed.

File: catlin/pkg/validator/task_validator_test.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+	res, err := parser.Parse()
+	assert.NilError(t, err)
+
+	v := ForKind(res)
+	result := v.Validate()
+
+	assert.Equal(t, 3, result.Errors)
+	assert.Equal(t, 3, len(result.Lints))
+	assert.Equal(t, `Step "hello" has an invalid image: could not parse reference: ubuntu`, result.Lints[0].Message)
```
Comment (vdemeester) [inline catlin/pkg/validator/task_validator_test.go:54]: wait, it should be the same error as `abc.io/fedora:latest` here. `ubuntu` is valid (because it's an "official" image), as it would become `docker.io/library/ubuntu` :upside_down_face:

Comment (SM43) [reply catlin/pkg/validator/task_validator_test.go:56]: `docker.io/library/ubuntu` would be valid but `ubuntu` invalid as per current implementation. Is that okay?

Comment (sthaha) [reply catlin/pkg/validator/task_validator_test.go:56]: @vdemeester `ubunbu` without a tag or a sha is valid?

Comment (sthaha) [reply catlin/pkg/validator/task_validator_test.go:56]: go-containerreg fails to parse reference to `ubuntu` - right? isn't that why we are getting this error?

Comment (SM43) [reply catlin/pkg/validator/task_validator_test.go:56]: `go-containerregistry/pkg/name.ParseReference` fails to validate the image ref as it does not have registry appended `docker.io`. Is there a way to which know is an official image?

Comment (sthaha) [reply catlin/pkg/validator/task_validator_test.go:56]: IIUC, https://github.com/google/go-containerregistry/blob/093adf2740a9617fbb6e6fe81ec37b00d3eb1b43/pkg/name/digest_test.go would fail validation of `ubuntu`

Comment (vdemeester) [reply catlin/pkg/validator/task_validator_test.go:56]: oh :joy_cat: so yeah, `ubuntu` should be valid. It is *always* better to use the full ref (`docker.io/library/ubuntu`) but `ubuntu` should be valid. An *official* image is like `ubuntu`, no username/group or host on top of it. `traefik`, `debian`, `fedora`, `mysql`, … all those are "official" images

Comment (vdemeester) [reply catlin/pkg/validator/task_validator_test.go:56]: We could *warn* in case of those, recommending using the full ref..

Comment (vdemeester) [reply catlin/pkg/validator/task_validator_test.go:56]: I am amazed `go-containerregistry` fails on those though…

Comment (vdemeester) [reply catlin/pkg/validator/task_validator_test.go:56]: Oh, I guess it depends on the `opt.strict` option thouhg…

Comment (SM43) [reply catlin/pkg/validator/task_validator_test.go:56]: Yeah. If we use `WeakValidation` then it doesn't fail for ubuntu but also for any image name for eg. `api` or `hub`
How do we differentiate between official and non-official images or should we warn for both?

Comment (vdemeester) [reply catlin/pkg/validator/task_validator_test.go:56]: So, I am fine using weak validation and have a `hub` image being seen as valid, even though it doesn't exists — as `foo.com/bar/baz:mytag` is gonna be seen as valid even though it doesn't exists.

And, yeah, I would warn for any image that is just *one part*, aka `ubuntu`, `api`, … being "official" or not, with a message "prefer using a fully qualified reference, like `library/ubuntu` or `docker.io/library/ubuntu` instead".

Comment (SM43) [reply catlin/pkg/validator/task_validator_test.go:56]: okay then Is `tag` required for all images?

Comment (vdemeester) [reply catlin/pkg/validator/task_validator_test.go:56]: a tag is required for all images and should be an error. images like `hub` or `ubuntu` should have a warn telling to use a fully qualified reference, like `library/hub`, `docker.io/library/ubuntu`.

File: catlin/pkg/validator/task_validator_test.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    image: api
+  - name: missing-repo
+    image: api:1.0
+  - name: no-tags
+    image: quay.io/api
+  - name: valid
+    image: registry.io/api:1.0
+`
+
+func TestTaskValidator_InvalidImageRef(t *testing.T) {
```
Comment (sthaha) [inline catlin/pkg/validator/task_validator_test.go:48]: @vdemeester @chmouel Could you please `lgtm` if this test covers all/most cases? 
To me this seems to cover all cases but will wait for you to do the honours :)

Comment (vdemeester) [reply catlin/pkg/validator/task_validator_test.go:48]: imho, this is too involved :sweat_smile: 
- I would rather not do any whitelist at all, and warn in all the case where a *one-word* image is used (be it `ubuntu`, `alpine` or `foobar`.
- according to the spec, nothing should make `quay.io/api` invalid. We **know** it is, because we know quay, but a program can't really make that assumption (and if it does, it feels weird). An exemple with https://r.j3ss.co/, the image references there are `r.j3ss.co/clisp:latest` and should be considered as valid.

File: catlin/pkg/validator/task_validator_test.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+	// whitelised image produces warning
+	assert.Equal(t, Warning, lints[0].Kind)
+	assert.Equal(t, `Step "whitelisted" uses image "ubuntu"; consider using a fully qualified name - e.g. docker.io/library/ubuntu`, lints[0].Message)
+
+	// image with variables produces warning
+	assert.Equal(t, Warning, lints[1].Kind)
+	assert.Equal(t, `Step "variables" uses image "ubuntu:$(params.img_ver)" that contains variables; skipping validation`, lints[1].Message)
+
+	// image with variables produces warning
```
Comment (vdemeester) [inline catlin/pkg/validator/task_validator_test.go:71]: that comment is not right as it is errors :stuck_out_tongue_closed_eyes:

**Comment by vdemeester** [comment]
Let me add some more validation examples
- `ubuntu` should be 1 warning (use full ref, aka docker.io/library/ubuntu) and 1 error (use a tag, different than latest)
- `ubuntu:latest` should be 1 warning (use full ref, aka docker.io/library/ubuntu) and 1 error (use a tag, different than latest)
- `foo` should be 1 warning (use full ref, aka docker.io/library/foo) and 1 error (use a tag, different than latest)
- `foo:bar` should be 1 warning (use full ref, aka docker.io/library/foo)
- `docker.io/foo` should be 1 error (use a tag, different than latest)
- `docker.io/foo:bar` should pass
- `r.j3ss.co/clisp` should be 1 error (use a tag, different than latest)
- `r.j3ss.co/clisp:latest` should be 1 error (use a tag, different than latest)
- `docker.io/vdemeester/foo` should be 1 error (use a tag, different than latest)
- `docker.io/vdemeester/foo:baz` should pass
- `quay.io/foo` should be 1 error (use a tag, different than latest)
- `quay.io/foo:bar` should pass
- `gcr.io/foo/bar/baz` should be 1 error (use a tag, different than latest)
- `gcr.io/foo/bar/baz:bar` should pass

There is more too
- `localhost:5000/foo:bar` should **probably** fail
- `172.16.3.4:3000/foo/bar:baz` should pass

File: catlin/pkg/validator/task_validator_test.go (hunk 9 above)
Comment (chmouel) [reply catlin/pkg/validator/task_validator_test.go:48]: 👍 on @vdemeester I dont think we should make it over complicated.

File: catlin/pkg/validator/validate_test.go [hunk 11]
```diff
@@ -63,33 +66,18 @@ metadata:
 spec:
   description: |-
     A summary of the resource
-
+
```
Comment (sthaha) [inline catlin/pkg/validator/validate_test.go:69]: nit: whitespace 🙏

**Review by sthaha — COMMENTED** [review]
Looks good to me but I will let @vdemeester and @chmouel review and merge.

File: catlin/go.mod [hunk 12]
```diff
@@ -3,16 +3,13 @@ module github.com/tektoncd/plumbing/catlin
 go 1.14
 
 require (
+	github.com/google/go-containerregistry v0.1.1
 	github.com/spf13/cobra v1.0.0
 	github.com/tektoncd/pipeline v0.15.2
 	go.uber.org/zap v1.15.0
 	gotest.tools/v3 v3.0.2
-
```
Comment (sthaha) [inline catlin/go.mod:10]: could you please revert back unrelated change?

File: catlin/pkg/validator/task_validator_test.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
+package validator
```
Comment (sthaha) [inline catlin/pkg/validator/task_validator_test.go:1]: license 🙏

File: catlin/go.mod (hunk 12 above)
Comment (SM43) [reply catlin/go.mod:10]: `go mod tidy` removed them I guess :thinking:

File: catlin/pkg/cmd/validate/validate.go [hunk 14]
```diff
@@ -84,10 +87,18 @@ func validate(cli app.CLI, path string) error {
...
 	out := cli.Stream().Out
 	for _, v := range result.Lints {
-		if v.Kind == validator.Error {
+		switch v.Kind {
+		case validator.Error:
 			fmt.Fprintf(out, "ERROR: %s\n", v.Message)
-		} else {
+		case validator.Warning:
+			fmt.Fprintf(out, "WARN : %s\n", v.Message)
+		case validator.Recommendation:
```
Comment (sthaha) [inline catlin/pkg/cmd/validate/validate.go:95]: unrelated to this patch: perhaps we should rename `Recommendation` to `Hint` :D

Comment (SM43) [reply catlin/pkg/cmd/validate/validate.go:95]: yeah :smiley_cat: will address in a separate patch

**Review by vdemeester — APPROVED** [review]
/meow
